src/Microsoft.AspNetCore.Dispatcher/Tree/TreeMatcher.cs

```diff
@@ -40,6 +37,7 @@ public TreeMatcher()
             }
 
             EnsureServicesInitialized(context);
+            Extensions.Primitives.ChangeToken.OnChange(() => ChangeToken, () => Volatile.Write(ref _dataInitialized, false));
```

Comment:
I tried to do this in the ctor but the `DataSource` is set after the ctor in `TreeMatcherFactory` and `ChangeToken` was null as a result

Reply:
You could do this in `EnsureServicesInitialized`. You can't really do it here, it's going to add another delegate every request.

Reply:
Can you add a using for this?